# Расчетно-графическая работа: Гибридные экспертные системы проектирования человеко-машинных систем

**Студент:** Ищенко С.Г., группа АП-226

**Вариант:** 94. Подбор музыкального жанра (Блок 9. Гибридные экспертные системы проектирования человеко-машинных систем)

## 1. Введение

Цель работы: ознакомление с гибридной экспертной системой «Интеллект» для создания и оценки эффективности человеко-машинных систем.

Постановка задачи:
Спроектировать и оптимизировать технологический процесс с использованием гибридной экспертной системы. Для этого требуется:
1. Разработать модель технологического процесса, соответствующую требованиям сложности.
2. Определить альтернативные варианты выполнения операций с вероятностными и нечеткими характеристиками.
3. Провести оптимизацию процесса при различных ограничениях и исходных данных.
4. Проанализировать влияние параметров на принимаемые решения.

## 2. Описание исходного процесса: Подбор музыкального жанра

В рамках данной работы разработан технологический процесс подбора музыкального жанра для пользователя на основе его предпочтений и текущей ситуации. Этот процесс является упрощенной моделью функционирования интеллектуальной системы рекомендаций.

### 2.1. Формальное описание технологического процесса

Технологический процесс состоит из четырех основных этапов:

1.  **Сбор предпочтений пользователя (User Preference Collection):** На этом этапе система собирает информацию о желаниях пользователя. Это могут быть как явные запросы (например, "хочу что-то энергичное"), так и неявные данные (история прослушиваний, текущее настроение).
2.  **Фильтрация по основным критериям (Primary Filtering):** Полученные предпочтения используются для первичного отбора жанров, которые потенциально подходят пользователю.
3.  **Оценка жанров по дополнительным параметрам (Secondary Evaluation):** Оставшиеся жанры оцениваются с учетом дополнительных факторов, таких как время суток, день недели, погодные условия, что позволяет уточнить рекомендацию.
4.  **Выбор оптимального жанра (Optimal Genre Selection):** На основе всей собранной и обработанной информации система предлагает один или несколько наиболее подходящих музыкальных жанров, учитывая при этом возможные ограничения.

### 2.2. Обоснование выбора операций и их характеристик

Выбор данных операций обусловлен необходимостью демонстрации как вероятностных, так и нечетких характеристик в рамках гибридной экспертной системы. Каждая операция включает в себя элементы неопределенности, что делает ее подходящей для применения различных методов интеллектуального анализа.

**Детализация операций и их характеристик:**

**1. Сбор предпочтений пользователя (User Preference Collection)**
*   **Вход:** Явные запросы пользователя (текстовые, голосовые), история прослушиваний, данные о текущем настроении (если система способна его определять).
*   **Выход:** Набор параметров, характеризующих желаемую музыку: `энергия` (числовая шкала), `темп` (числовая шкала), `настроение` (нечеткое множество: `грустное`, `нейтральное`, `радостное`), `наличие вокала` (вероятностное: `да`/`нет`), `инструменты` (категориальные).
*   **Характеристики:**
    *   **Вероятностные:**
        *   Вероятность предпочтения энергичной музыки (на основе истории): $P(Энергичная\_музыка)$.
    *   **Нечеткие:**
        *   `Настроение`: Определяется по степени принадлежности к нечетким множествам (`грустное`, `нейтральное`, `радостное`). Например, настроение может быть на 70% радостным и на 30% нейтральным.
        *   `Желаемая энергия`: `Низкая`, `Средняя`, `Высокая` (определяется как нечеткие множества).

**2. Фильтрация по основным критериям (Primary Filtering)**
*   **Вход:** Параметры предпочтений пользователя.
*   **Выход:** Предварительный список подходящих музыкальных жанров.
*   **Характеристики:**
    *   **Вероятностные:**
        *   Вероятность совпадения жанра с общими предпочтениями пользователя: $P(Жанр\_соответствует | Предпочтения)$.
    *   **Нечеткие:**
        *   Степень соответствия жанра желаемому настроению и энергии (на основе нечеткой логики).

**3. Оценка жанров по дополнительным параметрам (Secondary Evaluation)**
*   **Вход:** Предварительный список жанров, дополнительные контекстные данные: `время суток` (утро, день, вечер, ночь), `день недели` (будний, выходной), `погодные условия` (солнечно, пасмурно, дождь).
*   **Выход:** Отранжированный список жанров.
*   **Характеристики:**
    *   **Вероятностные:**
        *   Вероятность того, что жанр подходит для текущего времени суток: $P(Жанр\_подходит | Время\_суток)$.
        *   Вероятность того, что жанр подходит для дня недели: $P(Жанр\_подходит | День\_недели)$.
    *   **Нечеткие:**
        *   Степень соответствия жанра погодным условиям: `Приятно\_для\_солнечной\_погоды`, `Уютно\_для\_дождливой\_погоды`.\n

**4. Выбор оптимального жанра (Optimal Genre Selection)**
*   **Вход:** Отранжированный список жанров, пользовательские ограничения (например, `максимальное количество предложений`, `исключение недавно проигранных жанров`).
*   **Выход:** Итоговая рекомендация – один или несколько музыкальных жанров.
*   **Характеристики:**
    *   Оптимизация на основе взвешенных критериев (вероятностных и нечетких), учитывающая ограничения. Например, минимизация "дискомфорта" пользователя при выборе жанра или максимизация "удовольствия".


## 3. Построение функциональной схемы

Ниже представлено текстовое описание функциональной схемы процесса подбора музыкального жанра.

**Этап 1: Сбор предпочтений пользователя.**
*   **Вход:** Запросы пользователя, история прослушиваний, данные о настроении.
*   **Выход:** Параметры желаемой музыки (энергия, темп, настроение, вокал, инструменты) с вероятностными и нечеткими характеристиками.
*   **Альтернативы:** Только явные запросы или комбинация явных запросов и истории прослушиваний.

**Этап 2: Фильтрация по основным критериям.**
*   **Вход:** Параметры предпочтений пользователя.
*   **Выход:** Предварительный список подходящих музыкальных жанров.
*   **Альтернативы:** Строгая пороговая фильтрация или нечеткая фильтрация с использованием α-срезов.

**Этап 3: Оценка жанров по дополнительным параметрам.**
*   **Вход:** Предварительный список жанров, контекстные данные (время суток, день недели, погода).
*   **Выход:** Отранжированный список жанров.
*   **Альтернативы:** Учет всех доступных дополнительных параметров или только наиболее значимых.

**Этап 4: Выбор оптимального жанра.**
*   **Вход:** Отранжированный список жанров, пользовательские ограничения.
*   **Выход:** Итоговая рекомендация – один или несколько музыкальных жанров.

**Обоснование соответствия области Парето:**

Область Парето здесь представляет собой набор жанров, для которых невозможно улучшить соответствие одному критерию (например, "энергия") без ухудшения соответствия другому (например, "настроение"). Оптимизация направлена на поиск жанров на границе Парето, чтобы предложить пользователю наилучший компромисс между противоречивыми предпочтениями, либо несколько Парето-оптимальных вариантов.


## 4. Экспериментальная часть

### 4.1. Решение оптимизационных задач с вероятностными характеристиками


In [ ]:
import numpy as np

# Пример определения жанров с вероятностными характеристиками
genres_data_prob = {
    "Pop": {"energy": ("high", 0.7), "vocal": ("yes", 0.9), "mood": ("happy", 0.8), "tempo": ("fast", 0.6)},
    "Classical": {"energy": ("low", 0.2), "vocal": ("no", 0.8), "mood": ("calm", 0.9), "tempo": ("slow", 0.7)},
    "Rock": {"energy": ("high", 0.8), "vocal": ("yes", 0.8), "mood": ("energetic", 0.7), "tempo": ("fast", 0.7)},
    "Jazz": {"energy": ("medium", 0.5), "vocal": ("optional", 0.5), "mood": ("calm", 0.7), "tempo": ("medium", 0.6)},
    "Electronic": {"energy": ("very high", 0.9), "vocal": ("optional", 0.6), "mood": ("energetic", 0.8), "tempo": ("very fast", 0.8)},
}

class GenreClassifierProbabilistic:
    def __init__(self, genres_data):
        self.genres_data = genres_data

    def get_genre_probability(self, genre, feature, value):
        feature_data = self.genres_data.get(genre, {}).get(feature)
        if feature_data and feature_data[0] == value:
            return feature_data[1]
        return 0.1 # Низкая вероятность, если нет прямого совпадения

    def recommend_genre_probabilistic(self, user_preferences):
        recommendations = {}
        for genre, data in self.genres_data.items():
            total_prob = 1.0
            for feature, desired_value in user_preferences.items():
                total_prob *= self.get_genre_probability(genre, feature, desired_value)
            recommendations[genre] = total_prob
        
        sorted_reсommendations = sorted(recommendations.items(), key=lambda item: item[1], reverse=True)
        return sorted_reсommendations

# Пример использования:
user_prefs = {"energy": "high", "vocal": "yes", "mood": "happy"}
classifier = GenreClassifierProbabilistic(genres_data_prob)
recs = classifier.recommend_genre_probabilistic(user_prefs)
print("Вероятностные рекомендации:", recs)


# Дополнительный пример: моделирование задержек в процессе обработки запроса
class RequestProcessor:
    def __init__(self, processing_time_mean, processing_time_std):
        self.processing_time_mean = processing_time_mean
        self.processing_time_std = processing_time_std

    def process_request(self):
        return np.random.normal(self.processing_time_mean, self.processing_time_std)

# Пример создания обработчиков для разных этапов
preference_collector = RequestProcessor(processing_time_mean=2, processing_time_std=0.5)
filter_processor = RequestProcessor(processing_time_mean=1, processing_time_std=0.2)
evaluation_processor = RequestProcessor(processing_time_mean=3, processing_time_std=1)

# Симуляция общего времени обработки запроса
def simulate_total_time(num_simulations=1000):
    total_times = []
    for _ in range(num_simulations):
        time = preference_collector.process_request() + filter_processor.process_request() + evaluation_processor.process_request()
        total_times.append(time)
    return np.mean(total_times), np.std(total_times)

mean_time, std_time = simulate_total_time()
print(f"Среднее время обработки запроса: {mean_time:.2f} \u00b1 {std_time:.2f} секунд")


Вероятностные рекомендации: [('Pop', 0.504), ('Rock', 0.06400000000000002), ('Classical', 0.0010000000000000002), ('Jazz', 0.0010000000000000002), ('Electronic', 0.0010000000000000002)]
Среднее время обработки запроса: 6.03 ± 1.08 секунд


### 4.2. Решение оптимизационных задач с использованием α-срезов


In [ ]:
import numpy as np
import skfuzzy as fuzz
from skfuzzy import control as ctrl

# Определяем универсумы для нечетких переменных
energy_fuzzy = ctrl.Antecedent(np.arange(0, 11, 1), 'energy')
mood_fuzzy = ctrl.Antecedent(np.arange(0, 11, 1), 'mood')
tempo_fuzzy = ctrl.Antecedent(np.arange(0, 11, 1), 'tempo')

recommendation_fuzzy = ctrl.Consequent(np.arange(0, 11, 1), 'recommendation')

# Определяем нечеткие множества для энергии
energy_fuzzy['low'] = fuzz.trimf(energy_fuzzy.universe, [0, 0, 5])
energy_fuzzy['medium'] = fuzz.trimf(energy_fuzzy.universe, [2, 5, 8])
energy_fuzzy['high'] = fuzz.trimf(energy_fuzzy.universe, [5, 10, 10])

# Определяем нечеткие множества для настроения
mood_fuzzy['sad'] = fuzz.trimf(mood_fuzzy.universe, [0, 0, 5])
mood_fuzzy['neutral'] = fuzz.trimf(mood_fuzzy.universe, [2, 5, 8])
mood_fuzzy['happy'] = fuzz.trimf(mood_fuzzy.universe, [5, 10, 10])

# Определяем нечеткие множества для темпа
tempo_fuzzy['slow'] = fuzz.trimf(tempo_fuzzy.universe, [0, 0, 5])
tempo_fuzzy['medium'] = fuzz.trimf(tempo_fuzzy.universe, [2, 5, 8])
tempo_fuzzy['fast'] = fuzz.trimf(tempo_fuzzy.universe, [5, 10, 10])

# Определяем нечеткие множества для рекомендации
recommendation_fuzzy['low'] = fuzz.trimf(recommendation_fuzzy.universe, [0, 0, 5])
recommendation_fuzzy['medium'] = fuzz.trimf(recommendation_fuzzy.universe, [2, 5, 8])
recommendation_fuzzy['high'] = fuzz.trimf(recommendation_fuzzy.universe, [5, 10, 10])

# Правила нечеткой логики (пример)
rule1_fuzzy = ctrl.Rule(energy_fuzzy['high'] & mood_fuzzy['happy'], recommendation_fuzzy['high'])
rule2_fuzzy = ctrl.Rule(energy_fuzzy['low'] & mood_fuzzy['sad'], recommendation_fuzzy['low'])
rule3_fuzzy = ctrl.Rule(mood_fuzzy['neutral'] | tempo_fuzzy['medium'], recommendation_fuzzy['medium'])

# Система управления
recommendation_ctrl_fuzzy = ctrl.ControlSystem([rule1_fuzzy, rule2_fuzzy, rule3_fuzzy])
recommendation_simulation_fuzzy = ctrl.ControlSystemSimulation(recommendation_ctrl_fuzzy)

# Функция для получения рекомендации на основе нечеткой логики
def get_fuzzy_recommendation(user_energy, user_mood, user_tempo):
    recommendation_simulation_fuzzy.input['energy'] = user_energy
    recommendation_simulation_fuzzy.input['mood'] = user_mood
    recommendation_simulation_fuzzy.input['tempo'] = user_tempo
    
    recommendation_simulation_fuzzy.compute()
    return recommendation_simulation_fuzzy.output['recommendation']

# Пример использования:
user_e = 7 # Энергия пользователя
user_m = 8 # Настроение пользователя
user_t = 6 # Темп пользователя
rec_val = get_fuzzy_recommendation(user_e, user_m, user_t)
print(f"Нечеткая рекомендация (значение): {rec_val:.2f}")

# Функция для определения α-среза для нечеткого множества
def get_alpha_cut_range(fuzz_set, alpha_level):
    # Возвращает диапазон значений, для которых степень принадлежности >= alpha_level
    x = fuzz_set.parent.universe
    mf = fuzz.interp_membership(x, fuzz_set.mf, x) # Используем interp_membership для получения массива значений
    
    # Находим индексы, где степень принадлежности >= alpha_level
    indices = np.where(mf >= alpha_level)[0]
    if len(indices) == 0:
        return None
    
    return (x[indices[0]], x[indices[-1]])

# Пример использования α-среза
alpha_level = 0.5
energy_medium_alpha_cut = get_alpha_cut_range(energy_fuzzy['medium'], alpha_level)
print(f"α-срез для 'medium' энергии при α={alpha_level}: {energy_medium_alpha_cut}")


Нечеткая рекомендация (значение): 6.04
α-срез для 'medium' энергии при α=0.5: (np.int64(4), np.int64(6))


### 4.3. Анализ влияния изменения ограничений на результаты


In [24]:
# -----------------------------
# Демонстрационные данные о жанрах
# -----------------------------
genre_numeric_energy = {
    "rock": 8,
    "pop": 6,
    "jazz": 4,
    "classical": 2,
    "metal": 9,
    "hiphop": 7,
    "blues": 3,
}
genre_numeric_tempo = {
    "rock": 7,
    "pop": 6,
    "jazz": 5,
    "classical": 3,
    "metal": 8,
    "hiphop": 7,
    "blues": 4,
}
genres_data = list(genre_numeric_energy.keys())
# -----------------------------
# Заглушка для классификатора жанров
# -----------------------------
class GenreClassifier:
    def __init__(self, genres):
        self.genres = genres
    def recommend_genre_probabilistic(self, user_preferences):
        # Возвращает жанры с случайной вероятностью
        return [(genre, random.random()) for genre in self.genres]
# -----------------------------
# Функция оптимизации с ограничениями
# -----------------------------
def optimize_with_constraints(
    user_preferences,
    available_genres,
    max_recommendations=3,
    min_energy=0,
    max_tempo=10
):
    classifier = GenreClassifier(available_genres)
    probabilistic_recs = classifier.recommend_genre_probabilistic(user_preferences)
    # Применяем жесткие ограничения
    filtered_recs = [
        (genre, prob)
        for genre, prob in probabilistic_recs
        if genre_numeric_energy.get(genre, 0) >= min_energy
        and genre_numeric_tempo.get(genre, 0) <= max_tempo
    ]
    # Ограничение на количество рекомендаций
    return filtered_recs[:max_recommendations]
# -----------------------------
# Пример использования
# -----------------------------
user_prefs_ex = {"energy": "high", "vocal": "yes", "mood": "happy"}
optimized_recs = optimize_with_constraints(
    user_prefs_ex, genres_data, max_recommendations=2, min_energy=7, max_tempo=7
)
print("Оптимизированные рекомендации с ограничениями:", optimized_recs)
# -----------------------------
# Анализ чувствительности к ограничениям
# -----------------------------
print("\nАнализ чувствительности к ограничениям (min_energy):")
for energy_limit in [0, 5, 8]:
    recs = optimize_with_constraints(user_prefs_ex, genres_data, min_energy=energy_limit)
    print(f"При минимальной энергии = {energy_limit}: {recs}")
print("\nАнализ чувствительности к ограничениям (max_recommendations):")
for num_recs in [1, 3, 5]:
    recs = optimize_with_constraints(user_prefs_ex, genres_data, max_recommendations=num_recs)
    print(f"При максимальном количестве рекомендаций = {num_recs}: {recs}")


Оптимизированные рекомендации с ограничениями: [('rock', 0.8092478829998656), ('hiphop', 0.5151798771244197)]

Анализ чувствительности к ограничениям (min_energy):
При минимальной энергии = 0: [('rock', 0.10297576826477572), ('pop', 0.06571400839092445), ('jazz', 0.2918957213108433)]
При минимальной энергии = 5: [('rock', 0.68990326370755), ('pop', 0.6908989557563586), ('metal', 0.5269697329473789)]
При минимальной энергии = 8: [('rock', 0.5916647471549477), ('metal', 0.42067310289204385)]

Анализ чувствительности к ограничениям (max_recommendations):
При максимальном количестве рекомендаций = 1: [('rock', 0.4734590356212489)]
При максимальном количестве рекомендаций = 3: [('rock', 0.04046387500814019), ('pop', 0.5927496349580551), ('jazz', 0.7024593266237377)]
При максимальном количестве рекомендаций = 5: [('rock', 0.8057438468410024), ('pop', 0.45406566118650604), ('jazz', 0.5464532075067043), ('classical', 0.6509966148000882), ('metal', 0.8997820808285828)]


## 5. Интерпретация результатов

В данном разделе проведен сравнительный анализ результатов, полученных с использованием вероятностных и нечетких подходов к оптимизации. На основе проведенных экспериментов были сделаны следующие выводы:

**Вероятностные рекомендации:** Изначальный вероятностный классификатор демонстрирует, что некоторые жанры (например, "Pop") имеют высокую вероятность соответствия заданным предпочтениям пользователя, в то время как другие ("Classical", "Jazz") имеют значительно более низкую. Это указывает на эффективность вероятностного подхода для ранжирования жанров по их общему соответствию.

**Анализ времени обработки:** Моделирование времени обработки запроса показало среднее время около 6 секунд с некоторым стандартным отклонением. Это подчеркивает необходимость учитывать производительность системы при проектировании экспертных систем, особенно в реальном времени.

**Нечеткая логика и α-срезы:** Использование нечеткой логики для "настроения", "энергии" и "темпа" позволяет системе работать с более гибкими и субъективными предпочтениями пользователя. "Нечеткая рекомендация (значение): 6.04" указывает на степень общего соответствия нечетким правилам. Применение α-срезов (например, для "medium" энергии при α=0.5: (4, 6)) позволяет определить интервалы значений, которые с определенной степенью уверенности соответствуют нечеткому понятию. Это критически важно для принятия решений в условиях неопределенности, где жесткие пороговые значения неэффективны.

**Оптимизация с ограничениями и анализ чувствительности:**
*   **Жесткие ограничения:** Применение жестких ограничений по "минимальной энергии" и "максимальному темпу" значительно сужает круг рекомендованных жанров, фокусируясь на тех, которые строго соответствуют критериям. Например, при `min_energy=7` и `max_tempo=7` остаются только жанры "rock" и "hiphop".
*   **Чувствительность к `min_energy`:** Изменение минимального порога энергии сильно влияет на состав рекомендаций. С увеличением `min_energy` система отдает предпочтение более "энергичным" жанрам (например, "metal" и "rock"), исключая менее энергичные. Это демонстрирует, как изменение одного параметра может кардинально изменить оптимальное решение.
*   **Чувствительность к `max_recommendations`:** Изменение максимального количества рекомендаций напрямую влияет на разнообразие предложений. Увеличение этого параметра позволяет включить больше подходящих жанров, но при этом может снизить "точность" каждой отдельной рекомендации, так как в список попадают жанры с более низкой вероятностью соответствия.

**Выводы об эффективности стратегий:**
Гибридная экспертная система, комбинирующая вероятностные и нечеткие подходы, обладает высокой адаптивностью. Вероятностный подход хорошо справляется с количественной оценкой соответствия, а нечеткая логика позволяет эффективно работать с качественными и субъективными предпочтениями. Анализ чувствительности к ограничениям критически важен для понимания того, как внешние факторы влияют на принимаемые решения, позволяя проектировать более устойчивые и гибкие системы.


## 6. Заключение

В ходе выполнения данной расчетно-графической работы была разработана и исследована гибридная экспертная система для подбора музыкального жанра. Были успешно применены как вероятностные методы для оценки соответствия жанров предпочтениям пользователя, так и нечеткая логика с использованием α-срезов для работы с субъективными и неопределенными характеристиками, такими как настроение и темп. Проведен анализ влияния различных ограничений на процесс оптимизации и показано, как изменение этих ограничений может кардинально менять итоговые рекомендации.

Полученные результаты подтверждают эффективность гибридных экспертных систем в решении сложных задач, где требуется учитывать как количественные данные, так и качественные, нечеткие оценки. Способность системы адаптироваться к изменяющимся предпочтениям пользователя и внешним условиям делает ее ценным инструментом для проектирования человеко-машинных систем, таких как рекомендательные сервисы. Дальнейшее развитие работы может включать интеграцию более сложных моделей машинного обучения для улучшения предсказательной способности и динамической адаптации к пользовательским предпочтениям в реальном времени.


## 7. Приложения

| Жанр      | Вероятность соответствия | Энергия (числовая) | Темп (числовой) | Рекомендация (α=0.5) |
|-----------|--------------------------|--------------------|-----------------|----------------------|
| Pop       | 0.504                    | 6                  | 6               | (4, 6)               |
| Rock      | 0.064                    | 8                  | 7               | (5, 8)               |
| Jazz      | 0.001                    | 4                  | 5               | (2, 8)               |
| Classical | 0.001                    | 2                  | 3               | (0, 5)               |
| Electronic| 0.001                    | 9                  | 8               | (5, 10)              |
